### PR TITLE
some structure and app integration

### PR DESCRIPTION
**Pull request by djole103**
# DESIGN
`smile_local`     : [MAIN] local photo capture or scripted runs
`smile_server`  : [MAIN] respond to `verify` and `train` requests
`preprocessor` : data loading, cleaning, normalizing, reduction
`model`             : the core functionality, wraps a classifier and provides workflow
`classifiers`       : different clf's and how they should be tuned, train, predict ect
`config`             : globals and configuration

# TODO
- [x] Get the `/verify` flow down *where am I receiving data? body?
- [x] Make sure new photos are transformed to match current data set, consistency on dimensions
- [x] Collect stats+results, what do tables look like? what data is valuable for report?
- [x] Argument passing from command line *what args go as arg vs config? automating perfomance comparisons? * this has been changed to data collection script, input things and run there for now
- [ ] `classifiers` has pointless wrappers and redundancy, create a base classifier class and try to extend/inherit from clf instance to avoid `self.clf.clf` or pointless wrappers
- [x] initialization func for server startup needs to pull all our data to train *db conn *can we cache model/weights?

### CODE REVIEW

File: src/smile_server.py [hunk 1]
```diff
@@ -13,35 +13,57 @@
...
-        # return identity to user
-        return
+        identity = self.model.verify(processed[0])
+        return identity
+
+    def train(self):
+        fields = self.get_fields()
+        img = self.read_img(fields)
+        model.train(img, fields['label'])
+        return 200
```
Comment (djole103) [inline src/smile_server.py]: @hkdsun @seeARMS confirm how requests to vision look and how responses to app should be structured
suggesting:
Requests `POST /verify?image=b64img` and `POST /train?img=b64img&label=string`
Responses `200 name` and `200` where name = [valid name] | `no_match`

File: src/smile_server.py [hunk 2]
```diff
@@ -13,35 +13,57 @@
...
 
     def do_POST(self):
+        print("Post request")
+        #TODO: need to parse endpoint, separate from params
```
Comment (hkdsun) [inline src/smile_server.py]: can you elaborate?

File: src/smile_server.py [hunk 3]
```diff
@@ -13,35 +13,70 @@
...
+        processed = self.preprocessor.resize(processed) 
+        model.train(processed, fields['label'])
+        return 200
+
 
     def do_POST(self):
-        if self.path == "/verify":
-            return self.verify_img()
+        print("POST request")
+        #TODO: need to parse endpoint, separate from params
```
Comment (hkdsun) [inline src/smile_server.py:60]: can you elaborate?

File: src/model.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            (X_train, 
+            X_test, \
+            y_train, \
+            y_test, \
+            target_names, \
+            H, \
+            W
+            ) = data
+
+            pca = RandomizedPCA(n_components=PCA_N_COMPONENTS, whiten=True).fit(X_train)
```
Comment (hkdsun) [inline src/model.py]: So this is the very meat of our "method" of face recognition right? This is what we train our network on

File: src/model.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            eigenfaces = pca.components_.reshape((PCA_N_COMPONENTS, H, W))
+            X_train_processed = pca.transform(X_train)
+            self.train(X_train_processed, y_train)
+            self.save_model()
```
Comment (hkdsun) [inline src/model.py]: Should all of this routine (L34-L48) happen only if we are not getting our data from the db? Would you store training data in the db? I'm not sure I fully understand

File: src/smile_server.py (hunk 3 above)
Comment (djole103) [reply src/smile_server.py:60]: this is no longer relevant, i changed to `self.path.startswith` and that works

File: src/model.py (hunk 4 above)
Comment (djole103) [reply src/model.py]: No, what this does is tries to reduce feature dimensionality. It tries to cover the variance with fewer dimensions and keeps only the more relevant/impactful ones. It's actually a preprocessing step but I'm not sure why I get an error when I try to put this work into a function and return a pca instance. The whitening does some type of contrasting to try to make things more linearly separable or so we can notice feature effects more easily.
https://en.wikipedia.org/wiki/Principal_component_analysis

File: src/model.py (hunk 5 above)
Comment (djole103) [reply src/model.py]: L34-L48 is using the data to train a fresh model, whether the datas from db or the lfw database[1]. This only occurs if it couldn't find a cached already trained model. I'm assuming both are loaded in the same format. This is just photos and labels that have been divided for training and testing. (Testing not even used here actually)
Storing a cached model in DB is also possible and we should do that. Then it would check db and local for cache, if not load data to train.

File: src/classifiers.py [hunk 6]
```diff
@@ -74,24 +74,23 @@ def train(self, x_train, y_train):
...
+        return self.clf.predict_proba(x_test)
 
 
     def cache_string(self):
-        return "MLP_cached"
+        return "MLP"
 
 ########################## ....
 ## CLASSIFIER SELECTION ##    .
 ##########################    v
```
Comment (djole103) [inline src/classifiers.py]: it's not underneath the arrow now smh @hkdsun
